spec/controllers/admin/locations_controller_spec.rb

```diff
@@ -0,0 +1,164 @@
+require "rails_helper"
+
+RSpec.describe Admin::LocationsController, type: :controller do
+
+  context "when current user is not admin" do
+    before {get :index}
+
+    it "flash success message" do
+      expect(flash[:danger]).to match(I18n.t "permission_denide")
+    end
+
+    it "redirect to home page" do
+      expect(response).to redirect_to root_path
+    end
+  end
+
+  context "when current user is admin" do
+    login_admin
+    before {get :index}
+
+    it "should have a current_user" do
+      expect(subject.current_user).to_not eq nil
+    end
+
+    describe "GET locations#index" do
+      let(:locations){FactoryBot.create_list :location, 5}
+
+      it "show list location" do
+        locations.each do |location|
+          location.name
+        end
+      end
+    end
+
+    describe "POST locations#create" do
+      let(:location){FactoryBot.attributes_for :location}
+      let(:invalid_location){FactoryBot.attributes_for :invalid_location}
+
+      context "valid attributes" do
+        it "create new location" do
+          expect{post :create, params: {location: location}}
+          .to change(Location, :count).by 1
+        end
+
+        it "flash success message" do
+          post :create, params: {location: location}
+          expect(flash[:success]).to match(I18n.t "add_location_success")
+        end
+
+        it "redirect to admin_locations_path" do
+          post :create, params: {location: location}
+          expect(response).to redirect_to admin_locations_path
+        end
+      end
+
+      context "invalid attributes" do
+        before {Location.create!(name: "Đà Nẵng")}
+
+        it "does not create new location" do
+          expect{post :create, params: {location: invalid_location}}
+          .to_not change(Location, :count)
+        end
+
+        it "flash danger message" do
+          post :create, params: {location: invalid_location}
+          expect(flash[:danger]).to match(I18n.t "err_location")
+        end
+
+        it "redirect to admin_locations_path" do
+          post :create, params: {location: invalid_location}
+          expect(response).to redirect_to admin_locations_path
+        end
+      end
+    end
+
+    describe "PUT locations#update" do
+      let(:location){FactoryBot.attributes_for :location}
+      let(:invalid_locations){FactoryBot.attributes_for :invalid_locations}
```

Comment:
2 line code này lặp ở line 36 - 37 , em tìm hiểu về `subject in rspec` để tái sử dụng code nhé

Reply:
Cái ở dòng 37 là invalid_location là em tạo 1 factory riêng với FactoryBot.
Còn ở dòng 78 là invalid_locations là 1 factory con của factory: location (em thêm chữ "s" để phân biệt).
![Screenshot from 2019-04-10 13-32-54](https://user-images.githubusercontent.com/18730269/55856667-2b867e80-5b95-11e9-8943-3da15562a6f9.png)

Reply:
ờ, t nhìn ko ra (dead)